include/vrv/layerelement.h

```diff
@@ -427,6 +428,35 @@ class LayerElement : public Object,
     virtual int CountElementsInUnison(
         const std::set<int> &firstChord, const std::set<int> &secondChord, data_STEMDIRECTION stemDirection);
 
+    /**
+     * The note locations w.r.t. each staff, implemented for note and chord
+     */
+    virtual MapOfNoteLocs CalcNoteLocations() { return {}; };
+
+    /**
+     * The dot locations w.r.t. each staff, implemented for note and chord
+     * Since dots for notes on staff lines can be shifted upwards or downwards, there are two choices: primary and
+     * secondary
+     */
+    virtual MapOfDotLocs CalcDotLocations(int layerCount, bool primary) { return {}; };
+
+    /**
+     * Calculate the optimal dot location for a note or chord
+     * Takes two layers into account in order to avoid collisions of dots between corresponding notes/chords
+     */
+    MapOfDotLocs CalcOptimalDotLocations();
+
+    /**
+     * Helper to count the number of dots
+     * This can be used as an indicator to choose between different sets of dots
+     */
+    static int GetDotCount(const MapOfDotLocs &dotLocations);
+
+    /**
+     * Helper to count the collisions between two sets of dots
+     */
+    static int GetCollisionCount(const MapOfDotLocs &dotLocs1, const MapOfDotLocs &dotLocs2);
+
```

Comment:
We always declare methods and then variable members. Do not mix then. Also, keep Functors together and Static methods together to.